astropy_healpix/core.py

```diff
@@ -77,10 +84,92 @@ def level_to_nside(level):
     nside : int
         The number of pixels on the side of one of the 12 'top-level' HEALPix tiles.
     """
+    level = np.asarray(level, dtype=np.int64)
+
     _validate_level(level)
     return 2 ** level
 
 
+def nside_to_level(nside):
+    """
+    Find the HEALPix level for a given nside. (this is given by log2(nside))
+
+    This function is the inverse of `level_to_nside`.
+
+    Parameters
+    ----------
+    nside : int
+        The number of pixels on the side of one of the 12 'top-level' HEALPix tiles.
+        Must be a power of two.
+
+    Returns
+    -------
+    level : int
+        The level of the HEALPix cells
+    """
+    nside = np.asarray(nside, dtype=np.int64)
+
+    _validate_nside(nside)
+    return np.log2(nside)
+
+
+def uniq_to_levelipix(uniq):
+    """
+	Convert a HEALPix cell uniq number to its (level, ipix) equivalent
+
+    A uniq number is a 64 bit integer equals to : ipix + 4*(4**level). Please read
+    this `paper <http://ivoa.net/documents/MOC/20140602/REC-MOC-1.0-20140602.pdf>`_
+    for more details about uniq numbers.
+
+    Parameters
+    ----------
+    uniq : int
+        The uniq number of a HEALPix cell.
+
+    Returns
+    -------
+    level, ipix: int, int
+        The level and index of the HEALPix cell computed from ``uniq``.
+    """
+    uniq = np.asarray(uniq, dtype=np.int64)
+
+    level = ((np.log2(uniq//4)) // 2)
+    level = level.astype(np.int64)
```

Comment:
Why do you call `astype(np.int64)` here? You already do above, and this line should always be a no-op and can be removed, no?

Reply:
Hmm, doesn't np.log2() return floats?

Reply:
Yes and that is why I recast the result to int. Maybe there is a log2_integer numpy function  ? or we can use bitwise operation like [on this page](https://graphics.stanford.edu/~seander/bithacks.html#IntegerLogObvious) to compute it.

Reply:
Sorry, I didn't realise this returned a float. Yes, we should make an int.
Numpy doesn't have a log2_integer function.

Either keep as-is, or move the whole function to C and (for now) expose it via a Cython wrapper, if the speed is needed here. Given that @lpsinger and @mreineck comment a lot about performance here, and that this function might be performance-critical e.g. for MOC, that might be the way to go.

Personally I'm OK to merge as-is and leave that as a TODO comment for later, but if you have time to work on this now, you might as well do that change already in this PR.